### DISCUSSION TITLE
ESP built-in Temperature Sensor

### DISCUSSION THREAD

**Discussion by lboue** [arendst/Tasmota#17768] [Q&A] [answered]
### Goal
Could you tell me how to read the internal temperature sensor and display it in Tasmota.

### Doc 
The ESP32-S2/C3/S3/C2 has a built-in temperature sensor to measure the chip’s internal temperature. The temperature sensor module contains an 8-bit Sigma-Delta ADC and a DAC to compensate for the temperature offset.

- [ESP built-in Temperature Sensor Introduction](https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/temp_sensor.html)
- [Temperature Sensor Example](https://github.com/espressif/esp-idf/tree/49551cc48c/examples/peripherals/temp_sensor)

**Comment by barbudor** [comment, marked as answer]
The ESP32 internal temperature is now disabled by default as it is useless
If you really want that, you can enable it with `SetOption146 1`

https://tasmota.github.io/docs/Commands/#setoption146

**Reply by lboue** [reply]
Thanks

**Reply by teixeluis** [reply]
@barbudor why do you say it is useless? For me I see it correlates well with variations of the ambient temperature  (with an offset which is the heating of the chip itself). That can be used to define alarms if the temperature is too high, which can signal something is very wrong with the place where the device is located..

**Reply by teixeluis** [reply]
![image](https://github.com/arendst/Tasmota/assets/51710382/0d27ee8e-3f37-46b5-b61c-e51ec2de639c)

(violet graph on top is the ESP32 internal sensor, below several temperature readings of other sensors)

**Reply by sfromis** [reply]
My experience is that it is useless to say what what the temperature is, right after power up I've seen different ESP32 suggest both fridge and unbearably hot temperatures in a room with a "normal living temperature".

**Reply by teixeluis** [reply]
As you can see in my graph there are occasional spikes. Maybe that is related to moments of more cpu load, causing more heating. But if you filter that out over time, you obtain usable data.

**Reply by barbudor** [reply]
> with an offset which is the heating of the chip itself

Because this offset will depends the CPU activity which you can't model so it's mostly random

**Reply by teixeluis** [reply]
Yes but that is to be expected. Still useful for thermal management for example.

**Reply by barbudor** [reply]
That's what it is intended for: checking that you ESP is not roasting
But don't pretend it is usefull for room temperature

**Reply by sfromis** [reply]
And seeing an ESP32 provide a lot **lower** temperature than room temperature did lower my trust in that number to be "pretty cold".

**Reply by teixeluis** [reply]
Personally never observed that with my devices

**Reply by sfromis** [reply]
I've only noticed the clearly too low value on one ESP32 (haven't looked much), but my broader experience (with devices not having much to do) is seeing rather varying deviations between devices in very similar circumstances (close to idle at room temperature).

**Reply by techman83** [reply]
Mine mostly reported that they should be on fire, so I disabled it before it was disabled by default. So far none of them are actually on fire :shrug:
